File: bnfc/C/Interpret.c
#include <stdio.h>
#include <stdlib.h>

#include "Parser.h"
#include "Printer.h"
#include "Absyn.h"

#include "Interpreter.h"

int main(int argc, char **argv)
{
    FILE *input;
    Com parse_tree;

    if (argc > 1) {
        input = fopen(argv[1], "r");
        if (!input) {
            fprintf(stderr, "Error opening input file.\n");
            exit(1);
        }
    } else {
        input = stdin;
    }
    parse_tree = pCom(input);
    if (parse_tree) {
        /* printf("\nParse Succesful!\n"); */
        /* printf("\n[Abstract Syntax]\n"); */
        /* printf("%s\n\n", showCom(parse_tree)); */
        /* printf("[Linearized Tree]\n"); */
        /* printf("%s\n\n", printCom(parse_tree)); */
        exec(parse_tree);
        printState();
        return 0;
    }

    return 1;
}

File: bnfc/Haskell/Interpret.hs
module Main where

import System.IO ( stdin, hGetContents )
import System.Environment ( getArgs, getProgName )
import System.Exit ( exitFailure, exitSuccess )

import LexWhile
import ParWhile
import PrintWhile
import AbsWhile
import ErrM
import Interpreter

type ParseFun a = [Token] -> Err a

myLLexer = myLexer

type Verbosity = Int

putStrV :: Verbosity -> String -> IO ()
putStrV v s = if v > 1 then putStrLn s else return ()

-- runFile :: (Print a, Show a) => Verbosity -> ParseFun a -> FilePath -> IO ()
runFile v p f = putStrLn f >> readFile f >>= run v p

-- run :: (Print a, Show a) => Verbosity -> ParseFun a -> String -> IO ()
run v p s = let ts = myLLexer s in case p ts of
           Bad s    -> do putStrLn "\nParse              Failed...\n"
                          putStrV v "Tokens:"
                          putStrV v $ show ts
                          putStrLn s
                          exitFailure
           Ok  tree -> do print $ execProgram tree
                          exitSuccess


showTree :: (Show a, Print a) => Int -> a -> IO ()
showTree v tree
 = do
      putStrV v $ "\n[Abstract Syntax]\n\n" ++ show tree
      putStrV v $ "\n[Linearized tree]\n\n" ++ printTree tree

usage :: IO ()
usage = do
  putStrLn $ unlines
    [ "usage: Call with one of the following argument combinations:"
    , "  --help          Display this help message."
    , "  (no arguments)  Exec stdin."
    , "  (files)         Exec content of files."
    ]
  exitFailure

main :: IO ()
main = do
  args <- getArgs
  case args of
    ["--help"] -> usage
    [] -> hGetContents stdin >>= run 2 pCom
    fs -> mapM_ (runFile 2 pCom) fs
